# The Indian Start-up Ecosytem

### Installations

In [327]:
# %pip install seaborn
# %pip install plotly --upgrade

### Importation

In [328]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import plotly.express as px
import re
import matplotlib.pyplot as plt

### Load Data

In [329]:
data_2018 =  pd.read_csv('data/startup_funding2018.csv')
data_2019 =  pd.read_csv('data/startup_funding2019.csv')
data_2020 =  pd.read_csv('data/startup_funding2020.csv')
data_2021 =  pd.read_csv('data/startup_funding2021.csv')

## Hypothesis and Questions


Ho = Funding is not dependent on industry/sector.

Ha = Funding is highly dependent on the industry/sector a company is in.




1. What industry received the most funding?
2. Does loaction matter if one receives funding or not?
3. In which year is funding the highest?
4. How many companies received funding above $10,000,000 in each industry?
5. For companies in the pre-seed stage of funding, how much is the average pre-seed amount, and what industry are they in?

## Exploratory Data Analysis

In [330]:
data_2018.columns

Index(['Company Name', 'Industry', 'Round/Series', 'Amount', 'Location',
       'About Company'],
      dtype='object')

Change 2018 columns to match column specified in the project

In [331]:
data_2018.rename(columns={
    'Company Name':'Company/Brand',
    'Industry':'Sector',
    'About Company':'What it does',
    'Amount':'Amount($)',
    'Location':'HeadQuarter',
    'Round/Series':'Stage',
}, inplace=True)

In [332]:
data_2018.columns

Index(['Company/Brand', 'Sector', 'Stage', 'Amount($)', 'HeadQuarter',
       'What it does'],
      dtype='object')

In [333]:
data_2018.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage    Amount($)                       HeadQuarter  \
0      Seed       250000       Bangalore, Karnataka, India   
1      Seed  ₹40,000,000        Mumbai, Maharashtra, India   
2  Series A  ₹65,000,000           Gurgaon, Haryana, India   
3     Angel      2000000       Noida, Uttar Pradesh, India   
4      Seed            —  Hyderabad, Andhra Pradesh, India   

                                        What it does  
0  TheCollegeFever is a hub for fun, fiesta and f...  
1  A startup which aggregates milk from dairy far...  
2          Leading Online Loans Marketplace in India  
3  PayMe India is an innovative FinTech organizat...  
4  Eunimart is a one stop solution for merchants ...

In [334]:
data_2018['Founders'] = np.nan
data_2018['Investor'] = np.nan
data_2018.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage    Amount($)                       HeadQuarter  \
0      Seed       250000       Bangalore, Karnataka, India   
1      Seed  ₹40,000,000        Mumbai, Maharashtra, India   
2  Series A  ₹65,000,000           Gurgaon, Haryana, India   
3     Angel      2000000       Noida, Uttar Pradesh, India   
4      Seed            —  Hyderabad, Andhra Pradesh, India   

                                        What it does  Founders  Investor  
0  TheCollegeFever is a hub for fun, fiesta and f...       NaN       NaN  
1  A startup which aggregates milk from dairy far...       NaN       NaN  
2          Leading Online Loans Marketplace in India       NaN       NaN  
3  PayMe India is an innovative FinTech organizat...       NaN       NaN  
4  Eunimart is a one stop solution for merchants ...       NaN       NaN

In [335]:
data_2018.columns

Index(['Company/Brand', 'Sector', 'Stage', 'Amount($)', 'HeadQuarter',
       'What it does', 'Founders', 'Investor'],
      dtype='object')

#### Dropping all Founded columns since it is not required for the project

In [336]:
data_2019.drop(columns='Founded', inplace=True)
data_2020.drop(columns='Founded', inplace=True)
data_2021.drop(columns='Founded', inplace=True)
data_2020.drop(columns='Unnamed: 9', inplace=True)


##### I want to track the year a company was funded to accurately answer my hypothesis question

In [337]:
data_2018['Year'] = '2018'
data_2019['Year'] = '2019'
data_2020['Year'] = '2020'
data_2021['Year'] = '2021'

In [338]:
data_2019.columns

Index(['Company/Brand', 'HeadQuarter', 'Sector', 'What it does', 'Founders',
       'Investor', 'Amount($)', 'Stage', 'Year'],
      dtype='object')

In [339]:
data_2020.columns

Index(['Company/Brand', 'HeadQuarter', 'Sector', 'What it does', 'Founders',
       'Investor', 'Amount($)', 'Stage', 'Year'],
      dtype='object')

In [340]:
data_2021.columns

Index(['Company/Brand', 'HeadQuarter', 'Sector', 'What it does', 'Founders',
       'Investor', 'Amount($)', 'Stage', 'Year'],
      dtype='object')

Merging all my differnt dataframes

In [341]:
dataframes = [data_2018, data_2019, data_2020, data_2021]

data = pd.concat(dataframes, ignore_index=True)
data.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage    Amount($)                       HeadQuarter  \
0      Seed       250000       Bangalore, Karnataka, India   
1      Seed  ₹40,000,000        Mumbai, Maharashtra, India   
2  Series A  ₹65,000,000           Gurgaon, Haryana, India   
3     Angel      2000000       Noida, Uttar Pradesh, India   
4      Seed            —  Hyderabad, Andhra Pradesh, India   

                                        What it does Founders Investor  Year  
0  TheCollegeFever is a hub for fun, fiesta and f...      NaN      NaN  2018  
1  A startup which aggregates milk from dairy far...      NaN      NaN  2018  
2          Leading Online Loans Marketplace in India      NaN      NaN  2018  
3  PayMe India is an innovative FinTech organizat...      NaN      NaN  2018  
4  Eunimart is a one stop solution for merchants ...      NaN      NaN  2018

In [342]:
data.describe()

Company/Brand   Sector Stage    Amount($) HeadQuarter  \
count           2879     2861  1941         2873        2765   
unique          2214      873    75          699         172   
top         BharatPe  FinTech  Seed  Undisclosed   Bangalore   
freq              10      173   606          298         764   

                                             What it does         Founders  \
count                                                2879             2334   
unique                                               2691             1980   
top     BYJU'S is an educational technology company th...  Byju Raveendran   
freq                                                    5                7   

                         Investor  Year  
count                        2253  2879  
unique                       1777     4  
top     Inflection Point Ventures  2021  
freq                           36  1209

In [343]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2879 entries, 0 to 2878
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Company/Brand  2879 non-null   object
 1   Sector         2861 non-null   object
 2   Stage          1941 non-null   object
 3   Amount($)      2873 non-null   object
 4   HeadQuarter    2765 non-null   object
 5   What it does   2879 non-null   object
 6   Founders       2334 non-null   object
 7   Investor       2253 non-null   object
 8   Year           2879 non-null   object
dtypes: object(9)
memory usage: 202.6+ KB


In [344]:
data.duplicated().sum()

23

In [345]:
data.drop_duplicates(inplace=True)

In [346]:
data.duplicated().sum()

0

In [347]:
data.isna().sum()

Company/Brand      0
Sector            18
Stage            929
Amount($)          6
HeadQuarter      114
What it does       0
Founders         544
Investor         624
Year               0
dtype: int64

#### Engineering a feature 'Original Currency' to track on what rows the conversion of the currencies should be done 

In [348]:
data['Amount($)'] = data['Amount($)'].str.strip()
data['Amount($)'] = data['Amount($)'].apply(lambda x: str(x).replace(',', ''))
data['Original Currency'] = data['Amount($)'].apply(lambda x: 'Rupees' if '₹' in x or ('$' not in x and len(x) > 8) else 'Dollars')
data.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage  Amount($)                       HeadQuarter  \
0      Seed     250000       Bangalore, Karnataka, India   
1      Seed  ₹40000000        Mumbai, Maharashtra, India   
2  Series A  ₹65000000           Gurgaon, Haryana, India   
3     Angel    2000000       Noida, Uttar Pradesh, India   
4      Seed          —  Hyderabad, Andhra Pradesh, India   

                                        What it does Founders Investor  Year  \
0  TheCollegeFever is a hub for fun, fiesta and f...      NaN      NaN  2018   
1  A startup which aggregates milk from dairy far...      NaN      NaN  2018   
2          Leading Online Loans Marketplace in India      NaN      NaN  2018   
3  PayMe India is an innovative FinTech organizat...      NaN      NaN  2018   
4  Eunimart is a one stop solution for merchants ...      NaN      NaN  2018   

  Original Currency  
0           Dollars  
1            Rupees  
2            Rupees  
3           Dollars  
4           Dollars

##### Now let's convert the values in the amount columns that were tracked as rupee to dollars and delete the 'original currency' column. It has served its purpose

##### Asumptions

1. Each value in the amount column if it has a symbol, belongs to that currency
2. If value has no symbol and length of (value - symbol) > 8, then that value is in Rupees
3. Conversion rate been used is middle of the year 30th June 2018 

In [349]:
def clean_amount():
    data['Amount($)'] = data['Amount($)'].astype(str)
    data['Amount($)'] = data['Amount($)'].apply(lambda x: str(x).replace('$', ""))
    data['Amount($)'] = data['Amount($)'].apply(lambda x: str(x).replace('₹', ''))
    data['Amount($)'] = data['Amount($)'].apply(lambda x: '—' if re.search('osed|n', x) else x)
    data['Amount($)'] = data['Amount($)'].apply(lambda x: str(x).replace('—', '0'))
    data['Amount($)'] = data['Amount($)'].apply(lambda x: str(x).replace(',', ""))
    data['Amount($)'] = data['Amount($)'].apply(lambda x: str(x).replace(' ', ""))
    data['Amount($)'] = pd.to_numeric(data['Amount($)'], errors='coerce')
    data['Amount($)'] = data['Amount($)'].astype(float)
    

clean_amount()
data.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage   Amount($)                       HeadQuarter  \
0      Seed    250000.0       Bangalore, Karnataka, India   
1      Seed  40000000.0        Mumbai, Maharashtra, India   
2  Series A  65000000.0           Gurgaon, Haryana, India   
3     Angel   2000000.0       Noida, Uttar Pradesh, India   
4      Seed         0.0  Hyderabad, Andhra Pradesh, India   

                                        What it does Founders Investor  Year  \
0  TheCollegeFever is a hub for fun, fiesta and f...      NaN      NaN  2018   
1  A startup which aggregates milk from dairy far...      NaN      NaN  2018   
2          Leading Online Loans Marketplace in India      NaN      NaN  2018   
3  PayMe India is an innovative FinTech organizat...      NaN      NaN  2018   
4  Eunimart is a one stop solution for merchants ...      NaN      NaN  2018   

  Original Currency  
0           Dollars  
1            Rupees  
2            Rupees  
3           Dollars  
4           Dollars

In [350]:
coerce_values = data['Amount($)'].isna().sum()
percentage = (coerce_values/data.shape[0]) * 100

print(f'{percentage:.2f}% of our data is missing in the Amount column')

#Not bad

0.28% of our data is missing in the Amount column


##### Now, we convert the values that were tracked as rupees to dollars

In [351]:
exchange_rate = 0.0146
my_condition = data['Original Currency'] == 'Rupees'
data.loc[my_condition, 'Amount($)'] = data.loc[my_condition, 'Amount($)'] * exchange_rate

data.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage  Amount($)                       HeadQuarter  \
0      Seed   250000.0       Bangalore, Karnataka, India   
1      Seed   584000.0        Mumbai, Maharashtra, India   
2  Series A   949000.0           Gurgaon, Haryana, India   
3     Angel  2000000.0       Noida, Uttar Pradesh, India   
4      Seed        0.0  Hyderabad, Andhra Pradesh, India   

                                        What it does Founders Investor  Year  \
0  TheCollegeFever is a hub for fun, fiesta and f...      NaN      NaN  2018   
1  A startup which aggregates milk from dairy far...      NaN      NaN  2018   
2          Leading Online Loans Marketplace in India      NaN      NaN  2018   
3  PayMe India is an innovative FinTech organizat...      NaN      NaN  2018   
4  Eunimart is a one stop solution for merchants ...      NaN      NaN  2018   

  Original Currency  
0           Dollars  
1            Rupees  
2            Rupees  
3           Dollars  
4           Dollars

#### Now lets drop the Original currency column

In [352]:
data.drop(columns='Original Currency', inplace=True)

In [353]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2856 entries, 0 to 2878
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company/Brand  2856 non-null   object 
 1   Sector         2838 non-null   object 
 2   Stage          1927 non-null   object 
 3   Amount($)      2848 non-null   float64
 4   HeadQuarter    2742 non-null   object 
 5   What it does   2856 non-null   object 
 6   Founders       2312 non-null   object 
 7   Investor       2232 non-null   object 
 8   Year           2856 non-null   object 
dtypes: float64(1), object(8)
memory usage: 223.1+ KB


#### Now to deal with the HeadQuarters with multiple locations

Goal

- Create a helper function to clean up the headquaters column

Assumption

1. The first place listed is the principal headquaters
2. Since the entire dataset is about India, India can be dropped as well


Note 

- We have 114 null headquaters




In [354]:
data['HeadQuarter'] = data['HeadQuarter'].apply(str)
data['HeadQuarter'].str.lower()

null_headquaters = data[data['HeadQuarter'] == 'nan']
null_count = null_headquaters.shape[0]
null_percentage = (null_count/data.shape[0]) * 100

print(f'Total nan values in the heaquaters column is {null_count}')
print(f'This is {null_percentage:.2f}% of the heaquaters data')

Total nan values in the heaquaters column is 114
This is 3.99% of the heaquaters data


Now I will convert the nan string to something like 'Headquaters unknown'. 

In [355]:
filter1 = data['HeadQuarter'] == 'nan'

data.loc[filter1, 'HeadQuarter'] = 'Headquaters unknown'

In [356]:
def remove_excess_headquaters(string):
    new = []
    for letter in string:
        new += letter
    if ',' in new:
        comma_index = new.index(',')
        del new[comma_index:len(string)]
    headquaters = ''.join(new)
    return headquaters
    
data['HeadQuarter'] = data['HeadQuarter'].apply(remove_excess_headquaters)
data.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

      Stage  Amount($) HeadQuarter  \
0      Seed   250000.0   Bangalore   
1      Seed   584000.0      Mumbai   
2  Series A   949000.0     Gurgaon   
3     Angel  2000000.0       Noida   
4      Seed        0.0   Hyderabad   

                                        What it does Founders Investor  Year  
0  TheCollegeFever is a hub for fun, fiesta and f...      NaN      NaN  2018  
1  A startup which aggregates milk from dairy far...      NaN      NaN  2018  
2          Leading Online Loans Marketplace in India      NaN      NaN  2018  
3  PayMe India is an innovative FinTech organizat...      NaN      NaN  2018  
4  Eunimart is a one stop solution for merchants ...      NaN      NaN  2018

In [357]:
data.isna().sum()

Company/Brand      0
Sector            18
Stage            929
Amount($)          8
HeadQuarter        0
What it does       0
Founders         544
Investor         624
Year               0
dtype: int64

### Clean up the Sector column

My Goal

1. Create a function that adds specific keywords or expression into my predefined list

Goals 
1. There are nan and '-' that should be handled first. 
2. The goal is to replace those values with what the company does, this way we have some keywords to use to gauge what sector the company is in
3. Changing industries that are just have consumer to 'Business Consumer upon further investigation(regex prurpose)



###### Lets first visualize the Sector column

In [358]:
data['Sector'].unique()

array(['Brand Marketing, Event Promotion, Marketing, Sponsorship, Ticketing',
       'Agriculture, Farming',
       'Credit, Financial Services, Lending, Marketplace',
       'Financial Services, FinTech',
       'E-Commerce Platforms, Retail, SaaS',
       'Cloud Infrastructure, PaaS, SaaS',
       'Internet, Leisure, Marketplace', 'Market Research',
       'Information Services, Information Technology', 'Mobile Payments',
       'B2B, Shoes', 'Internet',
       'Apps, Collaboration, Developer Platform, Enterprise Software, Messaging, Productivity Tools, Video Chat',
       'Food Delivery', 'Industrial Automation',
       'Automotive, Search Engine, Service Industry',
       'Finance, Internet, Travel',
       'Accounting, Business Information Systems, Business Travel, Finance, SaaS',
       'Artificial Intelligence, Product Search, SaaS, Service Industry, Software',
       'Internet of Things, Waste Management',
       'Air Transportation, Freight Service, Logistics, Marine Transport

Imputing all nan and  - column with what the company does so we can guess based on the keywords

In [359]:
data['Sector'] = data['Sector'].apply(str)
data['Sector'] = data['Sector'].str.lower()

condition_1 = (data['Sector'] == 'nan') | (data['Sector'] == '—')
data.loc[condition_1, 'Sector'] = data.loc[condition_1, 'What it does']
data.head()

Company/Brand                                             Sector  \
0  TheCollegeFever  brand marketing, event promotion, marketing, s...   
1  Happy Cow Dairy                               agriculture, farming   
2       MyLoanCare   credit, financial services, lending, marketplace   
3      PayMe India                        financial services, fintech   
4         Eunimart                 e-commerce platforms, retail, saas   

      Stage  Amount($) HeadQuarter  \
0      Seed   250000.0   Bangalore   
1      Seed   584000.0      Mumbai   
2  Series A   949000.0     Gurgaon   
3     Angel  2000000.0       Noida   
4      Seed        0.0   Hyderabad   

                                        What it does Founders Investor  Year  
0  TheCollegeFever is a hub for fun, fiesta and f...      NaN      NaN  2018  
1  A startup which aggregates milk from dairy far...      NaN      NaN  2018  
2          Leading Online Loans Marketplace in India      NaN      NaN  2018  
3  PayMe India is an innovative FinTech organizat...      NaN      NaN  2018  
4  Eunimart is a one stop solution for merchants ...      NaN      NaN  2018

- Dealing with the multiple sector column to make it a little bit more concise
- I will create a function and use the re module to search for specific key words for  a particluar industry/sector


In [360]:
# for a more efficient regex grouping convert some more brief sector names

data['Sector'] = data['Sector'].str.lower()
data['Sector'] = data['Sector'].apply(lambda x: 'business consumer' if x == 'consumer' else x)
data['Sector'] = data['Sector'].apply(lambda x: 'e-vehicle' if x == 'ev' else x)
data['Sector'] = data['Sector'].apply(lambda x: 'tech' if x == 'it' else x)

def sector_redistribution(sector):
    if re.search('bank|fintech|financ|crypto|account|credit|venture|crowd|blockchain|fund|lending|trading|wealth|insurance|insurance|remittance|money|wealth|equity|investment|mortgage|nft|payments', sector):
        return 'Finance'
    elif re.search('automotive|air transport|transport|logistics|vehicle|transportation|delivery|aviation|vehicles|tyre|fleet|wheels|aero|luxury car|mobility|aeorspace|wl & rac protection|micro-mobiity', sector):
        return 'Transport'
    elif re.search('clean energy|energy|oil &|oil|solar|electricity|environment', sector):
        return 'Energy'
    elif re.search('analytics|consulting|human|career|erp|advertising|advertisement|market research|business|entrepre|recruit|hr|working|sultancy|advisory|work|job|management|skill|legal|crm', sector):
        return 'Business Intelligence'
    elif re.search('intelligence|tech|iot|crypto|cloud|artificial|data|net|things|apps|droid|software|computer|mobile|3d printing|funding platform|applications|file|embedded systems|online portals|fraud detection|search engine|nanotechnology|security|ai|saas|it company|augmented reality|platform|drone|ar startup|aas|app|online|/|virtual|it startup|photonics',sector):
        return 'Technology'
    elif re.search('beauty|cosmetic|skincare|fashion|wear|cosmetics|textiles|eye|jewellery|cloth', sector):
        return 'Beauty & Fashion'
    elif re.search('decor|fitness|training|wellness|personal care|deisgn|craft|design|podcast|lifestyle|spiritual|matrimony|living|cultural', sector):
        return 'LifeStyle'
    elif re.search('edutech|education|learn|children|child|collaboration|edtech', sector):
        return 'Education'
    elif re.search('beverage|catering|cook|food', sector):
        return 'Food & Bev'
    elif re.search('sports|esports|game|ball|player|manchester', sector):
        return 'Sports'
    elif re.search('customer|service|hospital|tourism|events|weedding|travel|hosts|booking|wedding|qsr', sector):
        return 'Hospitality'
    elif re.search('2|trade|enterpise|commerce|business|commercial|consumer goods|Marketplace|business consumer|marketing|retail|market|store|furniture|wholesale|wine & spirits|multinational|e-|packaging|sales|tplace|warehouse|fm|product|merchandise|reatil', sector):
        return 'Commerce'
    elif re.search('dental|health|health insurance|medic|supplement|biopharma|veterinary|pharma|heathcare|nutrition|hygiene|care|sanitation|bio|cannabis|tobacco|sciences', sector):
        return 'Health'
    elif re.search('agri|biotechnology|Industrial|farming|fish|milk|diary', sector):
        return 'Agriculture'
    elif re.search('battery|manufacturing|electronics|industrial automation|aerospace|conductor|gaming|robotics|engineering|mechanical|appliance|automation|ev startup|startup laboratory|e-vehicle', sector):
        return 'Manufacturing'
    elif re.search('communities|smart cities|government|classifieds|community|water|defense|pollution|translation & localization|taxation|maritime', sector):
        return 'Government'
    elif re.search('media|dating|music|audio|gaming|creative|entertainment|broadcasting|video|blogging|content|celebrity|ott', sector):
        return 'Media & Entertainment'
    elif re.search('apartment|real estate|home|interior|construction|rental|housing|accomodation|hauz', sector):
        return 'Real Estate'
    elif re.search('telecom|news|escrow|publication', sector):
        return 'Telecommunocations'
    else:
        return sector
    

data['Sector'] = data['Sector'].apply(sector_redistribution)

In [361]:
data['Sector'].unique()

array(['Commerce', 'Agriculture', 'Finance', 'Technology',
       'Business Intelligence', 'Transport', 'Manufacturing',
       'Food & Bev', 'Education', 'Energy', 'LifeStyle', 'Hospitality',
       'Health', 'Sports', 'Beauty & Fashion', 'Real Estate',
       'Media & Entertainment', 'Government', 'Telecommunocations'],
      dtype=object)

In [362]:
data['Sector'].unique().shape

(19,)

### Clean up stage column

I will be using the common finance stages used in business with a few assumptions 

Assumptions

- Pre-Seed :  Pre-seed and non equity assistance 
- Angel : angel investors 
- Seed : Every stage that contains a digit is also in the seed stage
- Pre-Series : This refers to the stage between the seed round and the series A round
- Series A : Series A
- Series B : Series B
- Series C : Series C
- Later Stage : Any staged beyound C that is not grant or equity
- Grant : Grant
- Equity : Equity

In [363]:
data['Stage'].unique()

array(['Seed', 'Series A', 'Angel', 'Series B', 'Pre-Seed',
       'Private Equity', 'Venture - Series Unknown', 'Grant',
       'Debt Financing', 'Post-IPO Debt', 'Series H', 'Series C',
       'Series E', 'Corporate Round', 'Undisclosed',
       'https://docs.google.com/spreadsheets/d/1x9ziNeaz6auNChIHnMI8U6kS7knTr3byy_YBGfQaoUA/edit#gid=1861303593',
       'Series D', 'Secondary Market', 'Post-IPO Equity',
       'Non-equity Assistance', 'Funding Round', nan, 'Fresh funding',
       'Pre series A', 'Series G', 'Post series A', 'Seed funding',
       'Seed fund', 'Series F', 'Series B+', 'Seed round', 'Pre-series A',
       'Pre-seed', 'Pre-series', 'Debt', 'Pre-series C', 'Pre-series B',
       'Bridge', 'Series B2', 'Pre- series A', 'Edge', 'Pre-Series B',
       'Seed A', 'Series A-1', 'Seed Funding', 'Pre-seed Round',
       'Seed Round & Series A', 'Pre Series A', 'Pre seed Round',
       'Angel Round', 'Pre series A1', 'Series E2', 'Seed Round',
       'Bridge Round', 'Pre seed

In [364]:
data['Stage'] = data['Stage'].apply(str)
data['Stage'] = data['Stage'].str.lower()

seed_condition = data['Stage'] == 'seed'
data.loc[seed_condition, 'Stage'] = 'Seed'

In [365]:
data['Stage'] = data['Stage'].apply(str)
def categorize_stage(stage):
    if re.search('angel', stage):
        return 'Angel'
    elif re.search('seed stage|seed round|seed\+|\d|funding round|seed investment|seed fund', stage):
        return 'Seed'
    elif re.search('non-equity assistance|pre-seed|early', stage):
        return 'Pre-Seed'
    elif re.search('bridge|pre-series', stage):
        return 'Pre-Series'
    elif re.search('series a|seies a|pre-series b|seed a', stage):
        return 'Series A'
    elif re.search('series b|b3|b2|series b\+|mid series', stage):
        return 'Series B'
    elif re.search('series c', stage):
        return 'Series C'
    elif re.search('grant', stage):
        return 'Grant'
    elif re.search('pe|private equity|secondary market', stage):
        return 'Equity'
    elif re.search('post-ipo equity|b3|h|f2|e2|d1|f1|series f1|series g|series f|series e|series i|series d|edge|corporate round', stage):
        return 'Later Stage'
    elif re.search('debt', stage):
        return 'Debt Financing'
    elif re.search('\$|https|nan|unknown|undisclosed', stage):
        return 'Unknown Stage'
    else:
        return stage
    
data['Stage'] = data['Stage'].apply(categorize_stage)

In [366]:
data['Stage'].unique()

array(['Seed', 'Series A', 'Angel', 'Series B', 'Pre-Seed', 'Equity',
       'Unknown Stage', 'Grant', 'Debt Financing', 'Later Stage',
       'Series C', 'Pre-Series'], dtype=object)

Making sure all categorical columns are in string format

 - First we have to make sure that all Nan values are changed to human understandable word like Unknown, for the investor and Founder columns.
 - We will need this for analysis instead of just droping it
 - Remember that NaN values will become 'nan' strings 
   - I will apply the lower() method for easy manipulation

In [367]:
data.isna().sum()

Company/Brand      0
Sector             0
Stage              0
Amount($)          8
HeadQuarter        0
What it does       0
Founders         544
Investor         624
Year               0
dtype: int64

In [368]:
data['Founders'] = data['Founders'].apply(str)
data['Investor'] = data['Investor'].apply(str)
data['Founders'] = data['Founders'].str.lower()
data['Investor'] = data['Investor'].str.lower()

founder_filter = data['Founders'] == 'nan'
investor_filter = data['Investor'] == 'nan'

data.loc[founder_filter, 'Founders'] = 'Founder Unknown'
data.loc[investor_filter, 'Investor'] = 'Investor Unknown'

Now we will have no null values instead it will be unknown

1. Lets change the types of the company and What is does data type to string just to be safe
2. I will also change the case of founders back to tile case since they contain human names #proper_nouns  #respect

In [375]:
data['Company/Brand'] = data['Company/Brand'].apply(str)
data['What it does'] = data['What it does'].apply(str)
data['Investor'] = data['Investor'].str.title()

In [376]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2856 entries, 0 to 2878
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company/Brand  2856 non-null   object 
 1   Sector         2856 non-null   object 
 2   Stage          2856 non-null   object 
 3   Amount($)      2848 non-null   float64
 4   HeadQuarter    2856 non-null   object 
 5   What it does   2856 non-null   object 
 6   Founders       2856 non-null   object 
 7   Investor       2856 non-null   object 
 8   Year           2856 non-null   object 
dtypes: float64(1), object(8)
memory usage: 223.1+ KB


In [377]:
data.isna().sum()

Company/Brand    0
Sector           0
Stage            0
Amount($)        8
HeadQuarter      0
What it does     0
Founders         0
Investor         0
Year             0
dtype: int64

Now the only columns we have that are missing are the coerce values when converting the amount column into float. Neat

#### Let's do some testing
